src/runtime/substitution.cc

```diff
@@ -2472,6 +2761,194 @@ T GraphSearchHelper::generic_sequence_optimize(
   return return_value;
 }
 
+/**
+ * @brief Experimental. To be merged into generic_sequence_optimize(). Top level
+ * DP search procedure for Unity with the consideration of memory usage.
+ *
+ * @tparam T Returned type
+ * @param graph Pre-optimization PCG
+ * @param sink_node Sink node of the PCG
+ * @param output_shape ???
+ * @param input_shape ???
+ * @return T Optimal result
+ */
+template <typename T>
+T GraphSearchHelper::generic_sequence_optimize_with_memory(
```

Comment:
Can we break up this function a bit? It's also quite long

Reply:
Let's plan the refactor in the future.

Reply:
Let's get an issue created, assigned, and linked here and in the code

Reply:
https://github.com/flexflow/FlexFlow/issues/620